### PR TITLE
Multichain

### PR DESCRIPTION
**Pull request by ryardley**
Closes: #140 
Closes: #75 
Closes: #134 

- [x] Simplify [evm package](https://github.com/gnosisguild/enclave/tree/ry/multi-chain/packages/ciphernode/evm/src)
- [x] Can send unrecovered errors directly to the bus for logging or recovery with `bus.err(EnclaveErrorType::Evm, error)`
- [x] Carry chain_id internally
- [x] Enable multichain args
- [x] Enable multichain config - [here is an example](https://github.com/gnosisguild/enclave/blob/ry/multi-chain/tests/basic_integration/lib/ciphernode_config.yaml)
- [x] Match on chain_id when writing to chain
- [x] Integrated rust ABI interfaces with actual EVM contracts [here](https://github.com/gnosisguild/enclave/blob/ry/multi-chain/packages/ciphernode/evm/src/enclave_sol_reader.rs#L11-L15), [here](https://github.com/gnosisguild/enclave/blob/ry/multi-chain/packages/ciphernode/evm/src/ciphernode_registry_sol.rs#L14-L18) and [here](https://github.com/gnosisguild/enclave/blob/ry/multi-chain/packages/ciphernode/evm/src/registry_filter_sol.rs#L16-L20) as suggested by @auryn-macmillan 

We should do the following a separate PR
- Remove println + do logging
- Support both http and websocket and automatically determine which protocol is required for reading and writing to evm
- Add default configuration (once we have an actual testnet deployment)

## Summary by CodeRabbit

- **New Features**
	- Added `src_chain_id` field to several data structures for enhanced event context.
	- Introduced a centralized configuration management approach for launching ciphernodes.
	- New dependency on `config` crate for improved configuration handling.
	- New message types and improved error handling in various modules.
	- Enhanced functionality for interacting with Ethereum smart contracts through new structures and methods.

- **Bug Fixes**
	- Enhanced error handling in `get_committee` method of `DistanceSortition`.

- **Documentation**
	- Updated configuration files to reflect new structure and parameters.

- **Tests**
	- Updated tests to accommodate the new `src_chain_id` field in event structures.

- **Chores**
	- Refactored scripts for launching ciphernodes to utilize a configuration file.
	- Added new dependency on `anyhow` crate for improved error handling.

### CODE REVIEW

File: packages/ciphernode/evm/src/ciphernode_registry_sol.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    rpc::types::Filter,
+    sol,
+    sol_types::SolEvent,
+};
+use anyhow::Result;
+use enclave_core::{EnclaveEvent, EventBus};
+
+use crate::helpers::{self, create_readonly_provider, ReadonlyProvider};
+
+sol! {
```
Comment (auryn-macmillan) [inline packages/ciphernode/evm/src/ciphernode_registry_sol.rs:14]: Is there a way to inherit this directly from the Enclave repo so we don't duplicate code and can maintain a single source of truth?

File: packages/ciphernode/evm/src/enclave_sol_reader.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+use alloy::primitives::{LogData, B256};
+use alloy::{
+    eips::BlockNumberOrTag, primitives::Address, rpc::types::Filter, sol, sol_types::SolEvent,
+};
+use anyhow::Result;
+use enclave_core::{EnclaveEvent, EventBus};
+
+sol! {
+    #[derive(Debug)]
+    struct E3 {
```
Comment (auryn-macmillan) [inline packages/ciphernode/evm/src/enclave_sol_reader.rs:13]: Can we use the evm package as the source of truth for interfaces?

**Review by auryn-macmillan — DISMISSED** [review]
lgtm

File: packages/ciphernode/evm/src/ciphernode_registry_sol.rs (hunk 1 above)
Comment (ryardley) [reply packages/ciphernode/evm/src/ciphernode_registry_sol.rs:14]: I did look into it but have not really found a solution that looked workable. Really happy to hear suggestions.

Comment (ryardley) [reply packages/ciphernode/evm/src/ciphernode_registry_sol.rs:14]: Actually I think I have found a possibility will test

File: packages/ciphernode/core/src/events.rs [hunk 3]
```diff
@@ -331,6 +346,7 @@ pub struct CiphertextOutputPublished {
 pub struct PlaintextAggregated {
     pub e3_id: E3id,
     pub decrypted_output: Vec<u8>,
+    pub src_chain_id: u64,
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/core/src/events.rs:349]: wrong

File: packages/ciphernode/core/src/events.rs [hunk 4]
```diff
@@ -305,11 +320,11 @@
     pub threshold_m: usize,
     pub seed: Seed,
     pub params: Vec<u8>,
-    // threshold: usize, // TODO:
-    // computation_type: ??, // TODO:
-    // execution_model_type: ??, // TODO:
-    // input_deadline: ??, // TODO:
-    // availability_duration: ??, // TODO:
+    pub src_chain_id: u64, // threshold: usize, // TODO:
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/core/src/events.rs:323]: wrong

File: packages/ciphernode/aggregator/src/publickey_aggregator.rs [hunk 5]
```diff
@@ -35,6 +43,7 @@
     sortition: Addr<Sortition>,
     e3_id: E3id,
     state: PublicKeyAggregatorState,
+    src_chain_id: u64,
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/aggregator/src/publickey_aggregator.rs:46]: wrong

File: packages/ciphernode/sortition/src/distance.rs [hunk 6]
```diff
@@ -16,22 +17,22 @@ impl DistanceSortition {
...
-                (score, *address)
+                Ok((score, *address))
             })
-            .collect::<Vec<(BigInt, Address)>>();
+            .collect::<Result<Vec<_>>>()?;
 
         scores.sort_by(|a, b| a.0.cmp(&b.0));
         let result = scores[0..self.size].to_vec();
-        result
+        Ok(result)
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/sortition/src/distance.rs:36]: This is actually helpful!

File: packages/ciphernode/evm/src/ciphernode_registry_sol.rs [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        let provider = self.provider.clone();
+        let filter = Filter::new()
+            .address(self.contract_address)
+            .from_block(BlockNumberOrTag::Latest);
+
+        ctx.spawn(
+            async move { helpers::stream_from_evm(provider, filter, bus, extractor).await }
+                .into_actor(self),
+        );
+    }
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/evm/src/ciphernode_registry_sol.rs:143]: it doesn't return an error!

File: tests/basic_integration/lib/ciphernode_config.yaml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+chains:
+  - name: "hardhat"
+    rpc_url: "ws://localhost:8545"
```
[the remark this answers is not in the document]
Comment (ryardley) [reply tests/basic_integration/lib/ciphernode_config.yaml:3]: This is a fair suggestion for a feature later

File: packages/ciphernode/evm/src/ciphernode_registry_sol.rs [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/evm/src/ciphernode_registry_sol.rs:143]: doesnt return a result

File: packages/ciphernode/evm/src/registry_filter_sol.rs [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+impl Handler<EnclaveEvent> for RegistryFilterSolWriter {
+    type Result = ();
+    fn handle(&mut self, msg: EnclaveEvent, ctx: &mut Self::Context) -> Self::Result {
+        match msg {
+            EnclaveEvent::PublicKeyAggregated { data, .. } => {
+                // Only publish if the src and destination chains match
+                if self.provider.get_chain_id() == data.src_chain_id {
+                    ctx.notify(data);
+                }
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/evm/src/registry_filter_sol.rs:72]: No

**Comment by ryardley** [comment]
Conversations have been resolved most were around fixing up stuff tabled for future PRs.

File: packages/ciphernode/evm/src/registry_filter_sol.rs [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+use enclave_core::{
+    BusError, E3id, EnclaveErrorType, EnclaveEvent, EventBus, OrderedSet, PublicKeyAggregated,
+    Subscribe,
+};
+use std::sync::Arc;
+
+sol!(
+    #[sol(rpc)]
+    NaiveRegistryFilter,
+    "../../evm/artifacts/contracts/registry/NaiveRegistryFilter.sol/NaiveRegistryFilter.json"
```
Comment (auryn-macmillan) [inline packages/ciphernode/evm/src/registry_filter_sol.rs:19]: @ryardley, I don't think this needs to block merging, but I am keen to understand what is missing from `IRegistryFilter.sol` that required you to use`NaiveRegistryFilter.sol` here? I assume it's the `publishCommittee()` method. In which case, I think we should perhaps include this in `IRegistryFilter.sol`, may also imply restructuring it to make it more generic.

Perhaps something like this, essentially an extra `data` field that can be used to pass in whatever other additional data (proofs, for example) that a given registry may need.

```solidity
    function publishCommittee(
        uint256 e3Id,
        address[] memory nodes,
        bytes memory publicKey,
        bytes memory data
    ) external;
```

**Review by auryn-macmillan — APPROVED** [review]
lgtm

**Review by hmzakhalid — APPROVED** [review]

**Pull request merged**